services/cd-service/pkg/service/batch_test.go

```diff
@@ -0,0 +1,282 @@
+					Environment: "production",
+					Config:      config.EnvironmentConfig{Upstream: &config.EnvironmentConfigUpstream{Latest: true}},
+				},
+				&repository.CreateApplicationVersion{
+					Application: "test",
+					Manifests: map[string]string{
+						"production": "manifest",
+					},
+				},
+				&repository.CreateApplicationVersion{
+					Application: "test",
+					Manifests: map[string]string{
+						"production": "manifest2",
+					},
+				},
+			},
+			Test: func(t *testing.T, svc *BatchServer) {
+				ops := []*api.BatchAction {}
+				for i := uint64(1); i <= 100; i++ {	// test exactly 100 first
+					ops = append(ops, &api.BatchAction{
+						Action: &api.BatchAction_Deploy{
+							Deploy: &api.DeployRequest{
+								Environment:  "production",
+								Application:  "test",
+								Version:      2-i%2,  // alternate between 1 and 2
+								LockBehavior: api.LockBehavior_Fail,
+							},
+						},
+					})
+				}
+				_, err := svc.ProcessBatch(
+					context.Background(),
+					&api.BatchRequest{
+						Actions: ops,
+					},
+				)
+				if err != nil {
+					t.Fatal(err)
+				}
+				{
+					version, err := svc.Repository.State().GetEnvironmentApplicationVersion("production", "test")
+					if err != nil {
+						t.Fatal(err)
+					}
+					if version == nil || *version != 2 {
+						t.Errorf("unexpected version: expected 2, actual: %d", *version)
+					}
+				}
+				// add one more action
+				ops = append(ops, &api.BatchAction{
+					Action: &api.BatchAction_Deploy{
+						Deploy: &api.DeployRequest{
+							Environment:  "production",
+							Application:  "test",
+							Version:      1,
+							LockBehavior: api.LockBehavior_Fail,
+						},
+					},
+				})
+				_, err = svc.ProcessBatch(
+					context.Background(),
+					&api.BatchRequest{
+						Actions: ops,
+					},
+				)
+				if err == nil {
+					t.Fatal("expected an error but got none")
+				}
+				s, ok := status.FromError(err)
+				if !ok {
+					t.Fatalf("error is not a status error, got: %#v", err)
+				}
+				if s.Code() != codes.InvalidArgument {
+					t.Errorf("invalid error code: expected %q, actual: %q", codes.InvalidArgument.String(), s.Code().String())
+				}
+				if s.Message() != "too many actions. limit is 100" {
+					t.Errorf("invalid error message: expected %q, actual: %q", "too many actions. limit is 100", s.Message())
+				}
+			},
+		},
+	}
+	for _, tc := range tcs {
+		tc := tc
+		t.Run(tc.Name, func(t *testing.T) {
+			t.Parallel()
+			dir := t.TempDir()
+			remoteDir := path.Join(dir, "remote")
+			localDir := path.Join(dir, "local")
+			cmd := exec.Command("git", "init", "--bare", remoteDir)
+			cmd.Start()
+			cmd.Wait()
+			repo, err := repository.NewWait(
+				context.Background(),
+				repository.Config{
+					URL:            remoteDir,
+					Path:           localDir,
+					CommitterEmail: "kuberpult@freiheit.com",
+					CommitterName:  "kuberpult",
+				},
+			)
+			if err != nil {
+				t.Fatal(err)
+			}
```

Comment:
I think this block is identical for different tests, so it should be extracted into a function `setupRepositoryTest`